### PR TITLE
Teuchos: Add optional yaml-cpp parser

### PR DESCRIPTION
**Pull request by cwschilly**
@rppawlo 

## Motivation

This PR implements an optional YAML parser using yaml-cpp.

Many of the additions in `Teuchos_YamlParser.cpp` came from [this commit](https://github.com/trilinos/Trilinos/commit/be224b6525490fe6528e40863663485321c5b7fc#diff-1a07c3a28e61987acc3a8dbe9f2daceed416d90e150c05e0e35c246499308981)  (PR #1779), which removed yaml-cpp in 2017.

This PR has been tested using `panzer/mini-em` in the NexGenAnalytics fork of Trilinos in [this PR](https://github.com/NexGenAnalytics/Trilinos/pull/219). That PR will not be merged (as it alters MiniEM source code), but it serves as an example of how the new parser could function in that context.

## Usage

Build Trilinos with `-D TPL_ENABLE_yaml-cpp=ON` in order to set the `HAVE_TEUCHOSPARAMETERLIST_YAMLCPP` flag. This will automatically cause any yaml input for a `Teuchos::ParameterList` to be parsed using yaml-cpp. If the flag is not set, the current Teuchos YAML parser will be used as is.

## Testing

I have added a new unit test (`yamlcpp_parser`) to ensure that the requested functionality is present. Specifically, it ensures the parser can correctly parse this yaml:
```yaml
mycode:
  list_of_2d_arrays:
    - [[1,2,3], [4,5,6]]
    - [[7,8,9], [10,11,12]]
  ragged_array:
    - [1,2,3]
    - [1,2,3,4]
  line_continuation: [
    1,2,3,
    4,5,6
  ]
  # allow unicode comments: ±
```
I have also added tests that ensure the yaml parsing exceptions (that were present in the original commit above) work as expected.

### CODE REVIEW

File: packages/teuchos/core/src/CMakeLists.txt [hunk 1]
```diff
@@ -32,6 +32,7 @@ SET(HAVE_TEUCHOSCORE_CXX11 ON)
 SET(YAML_CPP_DEF 0)
 IF(${PACKAGE_NAME}_ENABLE_yaml-cpp)
  SET(YAML_CPP_DEF 1)
+ SET(HAVE_TEUCHOSCORE_YAMLCPP ON)
 ENDIF()
```
Comment (rppawlo) [inline packages/teuchos/core/src/CMakeLists.txt:36]: This yaml block can be moved into the top level CMakeLists.txt file in the TeuchosParameterList subpackage.

**Review by rppawlo — COMMENTED** [review]
Thanks @cwschilly ! This looks good! A few comments:
- Since the core library doesn't care about yaml, I think we can move the config file stuff out of the core subpackage and into the parameterlist subpackage.
- If the yaml-cpp TPL is enabled, does it replace all use of the internal yaml parser (for all types of parmaeters? If so it might be possible to make the dependency on the TeuchosParser subpackage be optional instead of required. This depends on whether that parser subpackage is requires for xml parsing. We still need to support that.
- Not for part of this commit, but in the TeuchoComm subpackage are some functions that parse the yaml file and then send the data over MPI to all the other processes. I believe this should just work, but we might want to add the extra test cases as another test there if yaml-cpp is enabled.

Nice job with this! I'm adding some more reviewers to this as well.

**Comment by rppawlo** [comment]
@sebrowne - not sure what happened with the autotester here. All the actual builds passed, but the framework python3 check failed. Should I just retest?

File: packages/teuchos/core/src/CMakeLists.txt [hunk 2]
```diff
@@ -32,6 +32,7 @@ SET(HAVE_TEUCHOSCORE_CXX11 ON)
 SET(YAML_CPP_DEF 0)
```
Comment (bartlettroscoe) [inline packages/teuchos/core/src/CMakeLists.txt:32]: Where is this variable `YAML_CPP_DEF` being used?  I don't see where it is being used anywhere else in this CMake code.

**Review by bartlettroscoe — COMMENTED** [review]
I agree with @rppawlo's review.  To summarize:

**1.** The optional dependency on `yaml-cpp` needs to be moved from [`packages/teuchos/core/cmake/Dependencies.cmake`](https://github.com/NexGenAnalytics/Trilinos/blob/teuchos-add-yamlcpp-parser/packages/teuchos/core/cmake/Dependencies.cmake) to [`packages/teuchos/parameterlist/cmake/Dependencies.cmake`](https://github.com/NexGenAnalytics/Trilinos/blob/teuchos-add-yamlcpp-parser/packages/teuchos/parameterlist/cmake/Dependencies.cmake).

**2.** Move the logic for setting `HAVE_TEUCHOSCORE_YAMLCPP` from `teuchos/core/CMakeLists.txt` to `teuchos/parameterlist/CMakeLists.txt` and rename it to `HAVE_TEUCHOSPARAMETERLIST_YAMLCPP`

**3.** Create the file `teuchos/parameterlist/cmake/Teuchos_ParameterList_config.h.in` and configure it in `teuchos/parameterlist/src/CMakeLists.txt`.  Then include the generated file `Teuchos_ParameterList_config.h` in the needed Teuchos ParameterList soruce files (see [here](https://tribits.org/doc/TribitsUsersGuide.html#packagedir-cmake-packagename-config-h-in)).

Please let me know if you have any questions about making these changes.

**Comment by sebrowne** [comment]
@rppawlo yes, RETEST is appropriate.  Those tests timing out indicates some kind of system issue in my opinion.

**Comment by bartlettroscoe** [comment]
FYI: Last PR iteration failed with another random test failure (this time jumbled output on different streams).

Adding `AT: RETEST`.

**Comment by hkthorn** [comment]
> FYI: Last PR iteration failed with another random test failure (this time jumbled output on different streams).
> 
> SubProject	Name	Status	Time Status	Time	Proc Time	Details	Labels	Summary	Processors
> Belos	Belos_Tpetra_resolve_cg_hb_MPI_4	Failed	Passed	990ms	3s 960ms	Completed (Failed)	Belos	Unstable	4
> showing:
> 
> ```
> End ResuNumber of iterations performed for this solve (new solver): 872
> Number of iterations performed for this solve (new solver): 872
> Number of iterations performed for this solve (new solver): 872
> lt: TEST PASSED
> ```
> 
> (which breaks up `End Result: TEST PASSED` which is being looked for.)
> 
> Adding `AT: RETEST`.

A fix for this test has been merged this morning.

File: packages/teuchos/parameterlist/test/yaml/YamlParameterList.cpp [hunk 3]
```diff
@@ -309,5 +309,57 @@ namespace TeuchosTests
     TEST_EQUALITY(sublist.name(), "mycode->sublist");
   }
 
-} //namespace TeuchosTests
+#ifdef HAVE_TEUCHOSCORE_YAMLCPP
```
[the remark this answers is not in the document]
Comment (cwschilly) [reply packages/teuchos/parameterlist/test/yaml/YamlParameterList.cpp:312]: Yes; this YAML code was part of the "wishlist" we received with requested features of the new parser. I tested this block with/without yaml-cpp in the NGA fork (see the table at the top of [this PR](https://github.com/NexGenAnalytics/Trilinos/pull/219)).

File: packages/teuchos/core/src/CMakeLists.txt (hunk 2 above)
Comment (cwschilly) [reply packages/teuchos/core/src/CMakeLists.txt:32]: I don't see it used anywhere in Trilinos--it is likely a vestige from when yaml-cpp was used before. I will remove it in my next commit.

File: packages/teuchos/core/cmake/TeuchosCore_config.h.in [hunk 4]
```diff
@@ -11,6 +11,8 @@
 
 #cmakedefine HAVE_TEUCHOSCORE_QT
 
+#cmakedefine HAVE_TEUCHOSCORE_YAMLCPP
```
[the remark this answers is not in the document]
Comment (cwschilly) [reply packages/teuchos/core/cmake/TeuchosCore_config.h.in]: Fixed

File: packages/teuchos/core/src/CMakeLists.txt (hunk 1 above)
Comment (cwschilly) [reply packages/teuchos/core/src/CMakeLists.txt]: Fixed

File: packages/teuchos/parameterlist/src/Teuchos_YamlParser.cpp [hunk 5]
```diff
@@ -55,10 +55,159 @@
 #include "Teuchos_TwoDArray.hpp"
 
 #include "Teuchos_Reader.hpp"
+
+#ifdef HAVE_TEUCHOSCORE_YAMLCPP
```
[the remark this answers is not in the document]
Comment (cwschilly) [reply packages/teuchos/parameterlist/src/Teuchos_YamlParser.cpp]: Fixed (I chose `HAVE_TEUCHOSPARAMETERLIST_YAMLCPP`)

File: packages/teuchos/parameterlist/src/Teuchos_YamlParser_decl.hpp [hunk 6]
```diff
@@ -140,13 +141,33 @@ and looks better in YAML itself.
...
+{ \
+  public: \
+    Name(const std::string& arg) : ExceptionBase(arg) {} \
+};
+
+MAKE_EXCEPTION_TYPE(YamlKeyError)
+MAKE_EXCEPTION_TYPE(YamlScalarError)
+MAKE_EXCEPTION_TYPE(YamlSequenceError)
+MAKE_EXCEPTION_TYPE(YamlStructureError)
+MAKE_EXCEPTION_TYPE(YamlUndefinedNodeError)
```
[the remark this answers is not in the document]
Comment (cwschilly) [reply packages/teuchos/parameterlist/src/Teuchos_YamlParser_decl.hpp]: Fixed

File: packages/teuchos/core/src/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (cwschilly) [reply packages/teuchos/core/src/CMakeLists.txt]: Fixed

File: packages/teuchos/parameterlist/src/Teuchos_YamlParser.cpp [hunk 7]
```diff
@@ -1076,47 +1231,279 @@ namespace YAMLParameterList
...
+        }
+        else if(rawString == "false")
+        {
+          safe_set_entry<bool>(parent, key, false);
+        }
+        else
+        {
+          safe_set_entry<std::string>(parent, key, rawString);
+        }
+      }
```
Comment (cwschilly) [inline packages/teuchos/parameterlist/src/Teuchos_YamlParser.cpp:1350]: There used to be an extra `catch(...)` here that threw a `YamlScalarError` in the event that the provided value was not a int, double, bool, or string. However, the `if` statement at line 1323 guarantees that the given value must be a `::YAML::NodeType::Scalar` (ie an int, double, bool, or string) in order to enter this block, so that final catch statement would never be reached. As a result, I have removed that exception.

File: packages/teuchos/parameterlist/src/Teuchos_YamlParser.cpp [hunk 8]
```diff
@@ -1076,47 +1231,279 @@ namespace YAMLParameterList
...
+  }
+  else if(node.Type() == ::YAML::NodeType::Null)
+  {
+    // treat NULL as empty string (not an error)
+    safe_set_entry<std::string>(parent, key, std::string());
+  }
+  else
+  {
+    // Undefined
+    throw YamlUndefinedNodeError("Value type in a key-value pair must be one of: int, double, string, array, sublist.");
```
Comment (cwschilly) [inline packages/teuchos/parameterlist/src/Teuchos_YamlParser.cpp:1485]: I could not think of a way to test this exception--I have left a comment in the unit test

File: packages/teuchos/parameterlist/test/yaml/YamlParameterList.cpp [hunk 9]
```diff
@@ -104,14 +104,14 @@ namespace TeuchosTests
...
     RCP<ParameterList> params = Teuchos::getParametersFromYamlString(
         "My Awesome Problem:\n"
-        "\tMesh:\n"
-        "\t\tInline:\n"
-        "\t\t\tType: Quad\n"
-        "\t\t\tElements: [     10,     10 ]\n"
+        "  Mesh:\n"
+        "    Inline:\n"
+        "      Type: Quad\n"
+        "      Elements: [     10,     10 ]\n"
```
Comment (cwschilly) [inline packages/teuchos/parameterlist/test/yaml/YamlParameterList.cpp:114]: YAML does not accept tabs (only spaces), as documented [here](http://yaml.org/faq.html). Since yaml-cpp checks for proper yaml formatting before parsing, I had to change this unit test to use spaces instead of `\t`.

File: packages/teuchos/parameterlist/test/yaml/YamlParameterList.cpp [hunk 10]
```diff
@@ -281,8 +281,14 @@ namespace TeuchosTests
       " small number: 54\n"
       " big number: 72057594037927936\n");
     TEST_EQUALITY(pl->isType<int>("small number"), true);
+#ifdef HAVE_TEUCHOSPARAMETERLIST_YAMLCPP
+    TEST_EQUALITY(pl->isType<double>("big number"), true);
+    long long big_number = static_cast<long long>(pl->get<double>("big number"));
+    TEST_EQUALITY(big_number, 72057594037927936ll);
+#else
```
Comment (cwschilly) [inline packages/teuchos/parameterlist/test/yaml/YamlParameterList.cpp:288]: YAML only recognizes int, double, or string datatypes, meaning yaml-cpp has to parse this large number as a `double` instead of a `long long`.

**Comment by cwschilly** [comment]
@rppawlo I've updated the PR to make the requested changes. I've also run the ParameterList tests in TeuchosComm and everything works with using the yaml-cpp parser.

**Comment by cwschilly** [comment]
@rppawlo @bartlettroscoe @hkthorn If everything looks okay, could we run this PR through the autotester again?

File: packages/teuchos/parameterlist/test/yaml/YamlParameterList.cpp (hunk 10 above)
Comment (rppawlo) [reply packages/teuchos/parameterlist/test/yaml/YamlParameterList.cpp:288]: This comment is the only thing that makes me a uneasy. The parameter list supports "long long int" and from the looks of this test, our internal yaml parser could handle 64bit ints (maybe writing an illegal yaml file?). But now moving to yaml-cpp means we can no longer support "long long" data types directory when reading and writing to yaml. We might want to look into a better way to read and write long long entries to yaml so that the plist generated from a yaml file has the correct data type of long long instead of forcing everyone to use double.

File: packages/teuchos/parameterlist/test/yaml/YamlParameterList.cpp (hunk 9 above)
Comment (rppawlo) [reply packages/teuchos/parameterlist/test/yaml/YamlParameterList.cpp:114]: According the the official standard, I believe it does support tabs:

https://yaml.org/spec/1.2.2/

See section 5.5 on whitespace characters. The link specified above looks like outdated material (the copyright is 2006). What happens with yaml cpp?

**Review by bartlettroscoe — COMMENTED** [review]
I did not review every line of code carefully but it looks like the newer comments addressed all of mine and @rppawlo's suggestions summarized [above](https://github.com/trilinos/Trilinos/pull/12599#pullrequestreview-1793280262).  (I just wish we had coverage turned on so that we could see what lines of code were not being covered in automated testing.)

But it just occurred to me that we might consider changing the name of the external package/TPL `yaml-cpp` to remove the char `-` so that the automatically TriBITS-generated 
 CMake variable name `HAVE_<PACKAGE_NAME_UC>_<UPSTREAM_PACKAGE_NAME_UC>` will also be a valid C/C++ macro name (see below).   Some ideas are `yamlcpp` or YamlCpp`? Or we can just leave things are they are (and allow package names to have the `-` char because that is fine for CMake var names).  See more discussion below.

File: packages/teuchos/parameterlist/test/yaml/YamlParameterList.cpp (hunk 9 above)
Comment (cwschilly) [reply packages/teuchos/parameterlist/test/yaml/YamlParameterList.cpp:114]: yaml-cpp throws an `illegal map value` error when it tries to parse `\t`.

Example:
```cpp
  TEUCHOS_UNIT_TEST(YAML, PR1805)
  {
    RCP<ParameterList> params = Teuchos::getParametersFromYamlString(
        "My Awesome Problem:\n"
        "\tMesh:\n"
        "    Inline:\n"
        "      Type: Quad\n"
        "      Elements: [     10,     10 ]\n"
        "...\n"
        );
  }
```
throws
```
yaml-cpp: error at line 2, column 6: illegal map value
```

File: packages/teuchos/parameterlist/CMakeLists.txt [hunk 11]
```diff
@@ -1,6 +1,10 @@
 
 TRIBITS_SUBPACKAGE(ParameterList)
 
+IF(TPL_ENABLE_yaml-cpp)
+  SET(HAVE_TEUCHOSPARAMETERLIST_YAMLCPP ON)
```
[the remark this answers is not in the document]
Comment (rppawlo) [reply packages/teuchos/parameterlist/CMakeLists.txt:5]: My suggestion would be to do the right long term thing which I think would be to rename the tpl to "yamlcpp". The TPL has very little use after removal from the plist years ago. The only actual use is for three snapshotted packages - kokkos-kernels, percept and krino. KK uses it for performance testing while percept and krino have it listed as an optional LIB dependence. It probably would not be that disruptive of a change.

File: packages/teuchos/parameterlist/test/yaml/YamlParameterList.cpp (hunk 9 above)
Comment (stmcgovern) [reply packages/teuchos/parameterlist/test/yaml/YamlParameterList.cpp:114]: From https://yaml.org/spec/1.2.2/ we have BOTH
A. 
"5.5. White Space Characters
YAML recognizes two white space characters: space and tab."
and B.
 "6.1. Indentation Spaces
In YAML , structure is determined by indentation. In general, indentation is defined as a zero or more characters at the start of a line.
To maintain portability, [tab](https://yaml.org/spec/1.2.2/#white-space-characters) characters must not be used in indentation, since different systems treat [tabs](https://yaml.org/spec/1.2.2/#white-space-characters) differently."

Section 6.1 makes it seem like the example here (with tabs being used as indentation of the nodes) is not correct yaml. Section 5.5 indeed says that tabs are "valid yaml" and they can be used quoted in scalars or as in section 6.2(Separation Spaces) and 6.3(Line Prefixes), but the tab use case in this test here is for indentation.

Comment (rppawlo) [reply packages/teuchos/parameterlist/test/yaml/YamlParameterList.cpp:114]: Did a little more research on this. In sections 6.1 and 6.2 we have these comments:

To maintain portability, [tab](https://yaml.org/spec/1.2.2/#white-space-characters) characters must not be used in indentation, since different systems treat [tabs](https://yaml.org/spec/1.2.2/#white-space-characters) differently. 

The amount of indentation is a [presentation detail](https://yaml.org/spec/1.2.2/#presenting-the-serialization-tree) and must not be used to convey [content](https://yaml.org/spec/1.2.2/#nodes) information.

Outside [indentation](https://yaml.org/spec/1.2.2/#indentation-spaces) and [scalar content](https://yaml.org/spec/1.2.2/#scalar), YAML uses [white space](https://yaml.org/spec/1.2.2/#white-space-characters) characters for separation between tokens within a line. Note that such [white space](https://yaml.org/spec/1.2.2/#white-space-characters) may safely include [tab](https://yaml.org/spec/1.2.2/#white-space-characters) characters.

Example 6.1 shows where tabs can be safely used. I think the tabs before the "Inline" are safe to use, not sure if its worth fixing. I'm fine with these changes as is.

**Review by rppawlo — DISMISSED** [review]

File: packages/teuchos/parameterlist/CMakeLists.txt (hunk 11 above)
[the remark this answers is not in the document]
Comment (sebrowne) [reply packages/teuchos/parameterlist/CMakeLists.txt:5]: Technically isn't this similar to superlu_dist?  And there we (maybe everybody?) just dropped the underscore as Roger suggests, so I'd be in that boat too.

[the remark this answers is not in the document]
Comment (stmcgovern) [reply packages/teuchos/parameterlist/CMakeLists.txt:5]: I agree with the idea of renaming the library. I suggest that be done in a subsequent PR.

**Comment by rppawlo** [comment]
@cwschilly and @stmcgovern - I ran this through the empire test suite and nothing is passing. There is a difference in how the parser interprets empty nodes. The internal parser sets these to be sublists, but the current yaml-cpp sets these to be empry strings. Let's talk about this today at the regular meeting (or earlier if you want - send me an email offline). There is an new issue in EMPIRE that shows this.

**Comment by cwschilly** [comment]
> @cwschilly and @stmcgovern - I ran this through the empire test suite and nothing is passing. There is a difference in how the parser interprets empty nodes. The internal parser sets these to be sublists, but the current yaml-cpp sets these to be empry strings. Let's talk about this today at the regular meeting (or earlier if you want - send me an email offline). There is an new issue in EMPIRE that shows this.

@rppawlo @stmcgovern The offending line is [1479](https://github.com/trilinos/Trilinos/blob/a125832a96f12faa0dbbdf73f3134bffc80b8a9e/packages/teuchos/parameterlist/src/Teuchos_YamlParser.cpp#L1479) of Teuchos_YamlParser.cpp. Should be fairly straightforward fix.

**Comment by bartlettroscoe** [comment]
> @rppawlo @stmcgovern The offending line is [1479](https://github.com/trilinos/Trilinos/blob/a125832a96f12faa0dbbdf73f3134bffc80b8a9e/packages/teuchos/parameterlist/src/Teuchos_YamlParser.cpp#L1479) of Teuchos_YamlParser.cpp. Should be a fairly straightforward fix.

Does the Teuchos test suite need to be updated to catch this use case?  It looks like all of the existing tests passed.

**Comment by rppawlo** [comment]
> > @rppawlo @stmcgovern The offending line is [1479](https://github.com/trilinos/Trilinos/blob/a125832a96f12faa0dbbdf73f3134bffc80b8a9e/packages/teuchos/parameterlist/src/Teuchos_YamlParser.cpp#L1479) of Teuchos_YamlParser.cpp. Should be a fairly straightforward fix.
> 
> Does the Teuchos test suite need to be updated to catch this use case? It looks like all of the existing tests passed.

yes - should be easy to add the test case to make sure the empty nodes are sublists.

**Comment by cwschilly** [comment]
@rppawlo @bartlettroscoe @stmcgovern I've pushed the changes we've discussed and updated the `Usage` section of the PR description. We can now parse a `long long` if it ends with either `ll` or `LL`. I also added a test to make sure null nodes are read in as empty sublists

**Comment by cwschilly** [comment]
@rppawlo The latest commit improves the `long long` parsing to follow the same structure as the in-house parser.

I also noticed that the in-house parser accepts "yes" and "no" strings as bools, so I've added that functionality along with a  corresponding unit test.

**Review by rppawlo — DISMISSED** [review]
Thanks @cwschilly !

**Comment by stmcgovern** [comment]
The test failure is unrelated. Another commit is required to address an issue with capitalization of true/false for booleans.

**Comment by rppawlo** [comment]
We should probably check the old parser to see what variations are acceptable: "true", "True" and "TRUE"? Definitely want to be consistent with the old parser.

**Review by rppawlo — APPROVED** [review]
